### PR TITLE
terraform-providers.scalr: init at 1.0.4

### PR DESCRIPTION
**Pull request by dylanmtaylor**
###### Description of changes

Adds the `scalr` Terraform provider to `terraform-providers` package set.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by zowoq** [comment]
Use the [update-provider](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/cluster/terraform-providers/update-provider) script.

File: pkgs/applications/networking/cluster/terraform-providers/providers.json [hunk 1]
```diff
@@ -972,6 +972,15 @@
     "spdx": "MPL-2.0",
     "vendorHash": "sha256-kh1wv7cuWCC1rP0WBQW95pFg53gZTakqGoMIDMDSmt0="
   },
+  "scalr": {
+    "hash": "sha256-you7ohHHLbTAzwkEiErPSgf8Ppuw+EoZ+CDfOi35J4w=",
+    "homepage": "https://registry.scalr.io/scalr/scalr",
```
Comment (zowoq) [inline pkgs/applications/networking/cluster/terraform-providers/providers.json:977]: We currently don't support providers that aren't on registry.terraform.io

Comment (dylanmtaylor) [reply pkgs/applications/networking/cluster/terraform-providers/providers.json:977]: Oh. How could such support be added?

Comment (zowoq) [reply pkgs/applications/networking/cluster/terraform-providers/providers.json:977]: Not very easily. The packaging, scripts/automation assume or hardcode registry.terraform.io in multiple places, also other registries all seem to be just different enough that that they each require special handling.

TBH I'm not enthusiastic about adding support for other registries, it is rather a maintenance burden, especially when they only seem to host a single provider.

cc @zimbatm

Comment (dylanmtaylor) [reply pkgs/applications/networking/cluster/terraform-providers/providers.json:977]: I'm playing around with the overrideAttrs. I am trying to see if I can add it as a very special provider in the existing framework without having to make a new terraform-provider-scalr package folder.

**Comment by dylanmtaylor** [comment]
I think it might even be better if we add a condition to the script to only update the ones with terraform.io as a substring of the homepage.

**Comment by zowoq** [comment]
> I think it might even be better if we add a condition to the script to only update the ones with terraform.io as a substring of the homepage.

No, we've previously removed a provider that was hosted on a different registry that didn't work with the update script. https://github.com/NixOS/nixpkgs/pull/155959

**Comment by dylanmtaylor** [comment]
> > I think it might even be better if we add a condition to the script to only update the ones with terraform.io as a substring of the homepage.
> 
> No, we've previously removed a provider that was hosted on a different registry that didn't work with the update script. #155959

I could work on adding a special scalr condition to that script if that would be acceptable? I can get the latest version from GitHub's release endpoint since it is hosted there.

**Comment by dylanmtaylor** [comment]
I'm trying to find a good way to get this provider supported as it's something I use daily - the company I work for uses Scalr because it is massively cost-advantageous over TFC's business pricing. I'd love if you can provide guidance on how to do so in the least intrusive manner.

**Comment by zowoq** [comment]
As I said above I'm not enthusiastic about adding it. I don't want to end up with a collection of registries that only publish a single provider as it will be a maintenance burden. 

We don't need to add it to nixpkgs for your company to use it, can do so via a flake or overlay.

**Comment by dylanmtaylor** [comment]
Sure, I get that it's possible to maintain custom config, but I'd love to contribute to Nix for everyone to enjoy. Scalr is a special case here because Scalr will almost certainly never be in the official registry due to the fact that they compete with Terraform Cloud. I don't even know if Hashicorp would allow them to publish their module to TFC, but even so, they probably host their own as a safety net so that Hashicorp couldn't remove it.

**Comment by dylanmtaylor** [comment]
Using https://api.github.com/repos/Scalr/terraform-provider-scalr/releases/latest - we can grab the tag attribute to get the version. Perhaps with jq.

**Comment by zowoq** [comment]
> Scalr is a special case here because Scalr will almost certainly never be in the official registry due to the fact that they compete with Terraform Cloud.

Every provider that can't or won't publish on the official registry is a "special case". Doesn't change the fact that each one we add is a maintenance burden and once we add one we have no grounds to reject others.

**Comment by dylanmtaylor** [comment]
> > Scalr is a special case here because Scalr will almost certainly never be in the official registry due to the fact that they compete with Terraform Cloud.
> 
> Every provider that can't or won't publish on the official registry is a "special case". Doesn't change the fact that each one we add is a maintenance burden and once we add one we have no grounds to reject others.

I've been experimenting with handling GitHub as a source, I think I have something that won't be completely unmaintainable. Let me finish this up and I'll post it for comments.

**Comment by dylanmtaylor** [comment]
@zowoq - I think this handles providers on GitHub in a generic enough way that as long as the release uses a semver, it should work correctly and allow us to have overrides in the update script itself.

**Comment by zowoq** [comment]
Sorry but what you've pushed doesn't change what I said above, this when it isn't going to be merged. Closing.

**Comment by dylanmtaylor** [comment]
> Sorry but what you've pushed doesn't change what I said above, this isn't going to be merged. Closing.

I think a lot of the concern you had was that it didn't work with the update script in the past, correct? This PR made it work with the update script without breaking existing functionality, retrieving the latest version. It'd also be possible to reuse this code for almost any repository that happens to be hosted on GitHub by simply adding a few values to the array in the shell script.

Because the update-packages script would be able to fetch all of the relevant values, it could update packages.json using the existing workflow, making it fairly easy for this provider to be maintained.

I'd love for you to reconsider.

**Comment by zowoq** [comment]
Currently we have 140 providers, 137 work as is and 3 require a single line fix.

This adds a bunch of complexity that will only be of immediate benefit to one or two providers and will quite likely be of little benefit long term while increasing the overall maintenance burden.

**Comment by dylanmtaylor** [comment]
> Currently we have 140 providers, 137 work as is and 3 require a single line fix.
> 
> This adds a bunch of complexity that will only be of immediate benefit to one or two providers and will quite likely be of little benefit long term while increasing the overall maintenance burden.

I completely understand where you are coming from. On the flip side, there are thousands of `default.nix` files with logic and workarounds that are only of benefit to one package. I'd hardly say that doesn't make it a bad idea to package them if someone is willing to maintain it (as I am willing to maintain my code in this case), and if there are users of the package/provider that would benefit from it (and there are, evidenced by the fact that these packages have requests on nixpkgs). What if I were to break out the GitHub packages into a "terraform-providers-github" folder and have a separate script and workflow for them that I maintain and not touch your update script at all? We'd lose the benefit of having it all in one place, but I could maintain it independently and you wouldn't have to worry about bugs in my code breaking existing TFC packages.

**Comment by dylanmtaylor** [comment]
> This wouldn't actually change anything, they'd still need to use mkProvider so they can be in the `terraform-provider` namespace.
> 
> Offering to maintain it is appreciated but doesn't change that _I'd_ still have to maintain it regardless as couldn't make changes to one and not the other.

That's actually an extremely good point. So there'd be no benefits from splitting them out into different folders. I feel like I need to brainstorm on this some more. The burden of maintaining workarounds is largely due to conditional logic in the update script, right? What if we could somehow keep all the conditional logic for github packages separate and keep the existing logic intact, perhaps with some sort of wrapper that called common functions but overrode the few things that are special to github repo packages?

**Comment by zowoq** [comment]
> What if we could somehow keep all the conditional logic for github packages separate and keep the existing logic intact, perhaps with some sort of wrapper that called common functions but overrode the few things that are special to github repo packages?

My point of view is that adding providers that don't use the official registry is basically just doubling the complexity of the packaging and scripts, regardless of how simple it may be.

**Comment by zimbatm** [comment]
I don't view things exactly as zowoq. nixpkgs is here to provide utility, and Scalr is an important player in the Terraform ecosystem. Because we have a flat namespace here, there are some restrictions around name clashes, but we'll unlikely hit that with this provider.

Since we are building from source, it might be interesting to explore how we can scan the github tags on all the providers instead of querying the registry. That way it reduces the number of branches in the update code, and we could start supporting more registries.

@dylanmtaylor if the registry supports the same query methods as the official one, I'm happy to accept a PR at https://github.com/nix-community/nixpkgs-terraform-providers-bin

**Comment by dylanmtaylor** [comment]
> I don't view things exactly as zowoq. nixpkgs is here to provide utility, and Scalr is an important player in the Terraform ecosystem. Because we have a flat namespace here, there are some restrictions around name clashes, but we'll unlikely hit that with this provider.
> 
> Since we are building from source, it might be interesting to explore how we can scan the github tags on all the providers instead of querying the registry. That way it reduces the number of branches in the update code, and we could start supporting more registries.
> 
> @dylanmtaylor if the registry supports the same query methods as the official one, I'm happy to accept a PR at https://github.com/nix-community/nixpkgs-terraform-providers-bin

I love that you see that the Scalr provider would provide value here. Can we re-open this pull request so we can continue iterating on this idea? I don't know if it supports the same query methods since I just grabbed it from GitHub and built it that way, but it does work with the official Terraform client, so I imagine they at least partially reversed-engineered and implemented the TFC one. Perhaps you can help me with an investigation into that?

Thanks.

**Comment by zowoq** [comment]
> Can we re-open this pull request so we can continue iterating on this idea?

No, my opinion on this hasn't changed. One or two providers isn't worth the ongoing maintenance.

**Comment by zowoq** [comment]
> Since we are building from source, it might be interesting to explore how we can scan the github tags on all the providers instead of querying the registry. That way it reduces the number of branches in the update code, and we could start supporting more registries.

Honestly I think is is a bit of a policy change since the beginning of last year when we removed providers that weren't on the registry.

@zimbatm If you want to move forward with this idea do you want to take over as maintainer/codeowner?

**Comment by dylanmtaylor** [comment]
> > Can we re-open this pull request so we can continue iterating on this idea?
> 
> No, my opinion on this hasn't changed. One or two providers isn't worth the ongoing maintenance.

As @zimbatm said, Scalr is a major player in the Terraform ecosystem. I feel like the benefit to the community would be substantial and with further refinement, I bet we could reduce the amount of conditional logic such that it is even less than what I described in #222462. I'd like to open this PR back up, play around with some ideas (I would prefer to not open further PRs as it feels spammy here and the discussion can live in one place), and try to come to a solution that is more palatable. As I said earlier, I'm happy to be the maintainer of my changes.

**Comment by zowoq** [comment]
> As I said earlier, I'm happy to be the maintainer of my changes.

And as I said earlier, I'll end up having to maintain it regardless.

**Comment by dylanmtaylor** [comment]
So I just tried something - I messaged Scalr via one of my contacts there from work and asked if they would consider publishing to TFC. I imagine the answer might be no, as I don't think I'm the first one with this request, but it would eliminate the need for a workaround if they agree. I do understand your concern but I feel like the level of additional maintenance burden might be less than you are expecting. Regrettably, I don't feel like I will be able to convince you of this without trying to implement the solution proposed or one that we come up with together. If you're receptive to this idea, you can add me as a co-maintainer for the providers and I could help solution something that would work.

**Comment by dylanmtaylor** [comment]
I imagine that as established maintainers in the NixOS community, you have a large maintenance burden already, whereas I am fresh to the community and only have/want to have a handful of things that I maintain. Right now, this is just aws-workspaces and scalr-cli, but I am also hoping to take on gnome-extensions-cli and a couple of meraki packages as well as possibly chef-workstation. As they say in the development industry, I 'have the bandwidth' to take on (co-)maintainership of something like this Terraform provider script. I feel like what you have already is incredibly well designed, and adding the logic to pull the special providers from GitHub isn't a huge undertaking. I also have extensive experience with GitHub actions and wouldn't mind troubleshooting that. I'm happy to work alongside you on this. Let me know what you're receptive too here - I don't plan to just merge in my PR and disappear.

**Comment by zimbatm** [comment]
To add some historical context; Terraform introduced the registry around 0.11. We supported the old system and the registry for a while, leading to complicated code. Once 1.0 came out, zowoq helped a lot to clean things and remove the old packages that didn't migrate to the registry. That's when we agreed not to include packages outside the registry. And generally, to avoid packages with overrides and they create more maintenance burden. zowoq also set up GitHub Actions to automate the updates, helped a lot on the nixpkgs-terraform-providers-bin repo and also takes really good care of that part of nixpkgs.

@zowoq to me, the policy made sense in that context. I don't mind re-evaluating it, as long as the new policy is sustainable. At the same time, I also respect that you shouldn't be forced to maintain more code than you are comfortable with (this reminds me of the GHA discussions we were having as well). Ideally, we find a solution that satisfies you both.

Regarding the design, we hit a local maxima, but I think there is still room to improve things.

One would be to split the fat providers.json into one JSON per provider + a builtins.readDir to load them all. This makes it easier to work with individual providers, and also reduces the risk of merge conflicts. It also makes it easier to see all the available providers by navigating in nixpkgs.

Another one would be if we can use the ryantm updaters bot. See how each provider can expose an updateScript, so we don't have to maintain our own github actions script.

I think these two changes would make it easier to split the maintenance further down as well. What do you think? Making room for new contributors is also important.

**Comment by zowoq** [comment]
>  I messaged Scalr via one of my contacts there from work and asked if they would consider publishing to TFC. 

Until we have an answer about this I don't see that we need to discuss anything TBH, if it does get added to the registry then there really isn't any reason to make changes here.

That said I will comment on two things:

> One would be to split the fat providers.json into one JSON per provider + a builtins.readDir to load them all.

I've thought about this before. I've found make changes across all of the providers to be very easy with everything in one file, I'd be strongly opposed to splitting it up.

> See how each provider can expose an updateScript, so we don't have to maintain our own github actions script.

The action just runs all of the updateScripts at once.

**Pull request closed without merging**